blockprod/src/detail/mod.rs

```diff
@@ -328,13 +321,33 @@ impl BlockProduction {
         let (job_remover_func, end_confirm_receiver) = self.job_manager.make_job_stopper_function();
         let _job_remover = OnceDestructor::new(move || job_remover_func(job_key));
 
+        let mut previous_attempt = timestamp_at_start;
+        let mut previous_consensus_status =
+            self.chain_config.net_upgrade().consensus_status(tip_at_start.block_height());
+
         loop {
             let timestamp =
                 BlockTimestamp::from_duration_since_epoch(self.time_getter().get_time());
 
+            match previous_consensus_status {
+                RequiredConsensus::IgnoreConsensus | RequiredConsensus::PoW(_) => {}
+                RequiredConsensus::PoS(_) => {
+                    if previous_attempt == timestamp && timestamp != timestamp_at_start {
+                        sleep(Duration::from_secs(1)).await;
+                        continue;
+                    }
+                }
+            }
```

Comment:
If I understand correctly, this will result in producing blocks blocking forever until a block is found. I'm not particularly against this, but I think this can be engineered in a better way.

My concern here is that something is missing in this calculation, and the abstraction is not properly done. Consider the following: Let's call the current timestamp at start T0, and the maximum timestamp in the future at start T1, and the timestamp of the previous block TB. The first scan goes over the range `[TB,T1]`. But then... it seems that with this setting, we'll save `T0+t` all the time, where `t` is the time elapsed since we started staking the first time, and we'll "try to stake" again repeatedly over a certain range. Now the problem is that there's a range that's always re-checked in this mechanism, and this range is `[T0+t, T1+t]`, while in reality, we want to check only `[T1,T1+t]`, because `[TB,T1]` has already been covered. So this solution, while it works, it kind creates an efficiency issue.

Not that we have to fix this right now... but this has to be addressed sooner or later. The fix should have an object that takes all these considerations into account and updates itself such that this doesn't happen.

Reply:
Yep, thanks. As above, I've redone a few timestamp calculations and I think my latest change fixes your comment.

The issue I had was that because we are starting at Genesis for the timestamp, tests were failing because the block timestamps were too old. As a temporary solution, I've temporarily increased `ChainstateConfig.max_tip_age` to a large number so that I don't block Pavel moving forward.

I tried to increase `max_tip_age` for only the functional tests, but it was taking too long working out how the `ChainstateConfigFile` was built for the tests, so I'll leave it for later.

Reply:
I don't think this issue was addressed. Please add a comment about it. Feel free to discuss with me on Slack if it's not clear.

Reply:
Thanks for showing me how incorrect this was. I've now updated it with an `AtomicU64` so that we continue from where we left off when we failed to produce a block.

Hopefully the updated code fixes this issue.